mtenn/conversion_utils/visnet.py

```diff
@@ -0,0 +1,237 @@
+    from torch_geometric.nn.models import ViSNet as PygVisNet
+    HAS_VISNET = True
+except ImportError:
+    pass
+
+
+class EquivariantVecToScalar(torch.nn.Module):
+    # Wrapper for PygVisNet.EquivariantScalar to implement forward() method
+    def __init__(self, mean, reduce_op):
+        super(EquivariantVecToScalar, self).__init__()
+        self.mean = mean
+        self.reduce_op = reduce_op
+    def forward(self, x):
+        # dummy variable. all atoms from the same molecule and the same batch
+        batch = torch.zeros(x.shape[0], dtype=torch.int64, device=x.device)
+
+        y = scatter(x, batch, dim=0, reduce=self.reduce_op)
+        return y + self.mean
+
+
+if HAS_VISNET:
+    class ViSNet(torch.nn.Module):
+        def __init__(self, *args, model=None, **kwargs):
+            super().__init__()
+            ## If no model is passed, construct default ViSNet model, otherwise copy
+            ##  all parameters and weights over
+            if model is None:
+                self.visnet = PygVisNet(*args, **kwargs)
+            else:
+                atomref = model.prior_model.atomref.weight.detach().clone()
+                model_params = (
+                    model.lmax,
+                    model.vecnorm_type,
+                    model.trainable_vecnorm,
+                    model.num_heads,
+                    model.num_layers,
+                    model.hidden_channels,
+                    model.num_rbf,
+                    model.trainable_rbf,
+                    model.max_z,
+                    model.cutoff,
+                    model.max_num_neighbors,
+                    model.vertex,
+                    model.reduce_op,
+                    model.mean,
+                    model.std,
+                    model.derivative, # not used. originally calculates "force" from energy
+                    atomref,
+                )
+                self.visnet = PygVisNet(*model_params)
+                self.load_state_dict(model.state_dict())
+
+            self.readout = EquivariantVecToScalar(self.visnet.mean, self.visnet.reduce_op)
+
+        def forward(self, data):
+            """
+            Computes the energies or properties (forces) for a batch of
+            molecules.
+
+            Args:
+                z (torch.Tensor): The atomic numbers.
+                pos (torch.Tensor): The coordinates of the atoms.
+                batch (torch.Tensor): A batch vector,
+                    which assigns each node to a specific example.
+
+            Returns:
+                x (torch.Tensor): Scalar output based on node features and vector features.
+            """
+            pos = data["pos"]
+            z = data["z"]
+
+            # all atom in one pass from the same molecule
+            batch = torch.zeros(z.shape[0], device=z.device)
+            x, v = self.visnet.representation_model(z, pos, batch)
+            x = self.visnet.output_model.pre_reduce(x, v)
+            x = x * self.visnet.std
+
```

Comment:
following up on an in-person discussion, we want to add this code from the original ViSNet that got dropped in the transfer:
```suggestion

        if self.visnet.prior_model is not None:
            x = self.visnet.prior_model(x, z)

```